### PR TITLE
Reorganize optimize, and allow callables other than Functions as f g …

### PR DESCRIPTION
**Pull request by pkofod**
…or h.

### CODE REVIEW

File: src/optimize/multivariate/interface.jl [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+end
+
+function optimize{M <: Union{FirstOrderSolver, SecondOrderSolver}}(f, initial_x::AbstractArray, method::M, options::Options)
+    d = ifelse(M <: FirstOrderSolver, OnceDifferentiable(f, initial_x), TwiceDifferentiable(f, initial_x))
+    optimize(d, initial_x, method, options)
+end
+function optimize{M <: Union{FirstOrderSolver, SecondOrderSolver}}(f::Function, initial_x::AbstractArray, method::M, options::Options)
+    d = ifelse(M <: FirstOrderSolver, OnceDifferentiable(f, initial_x), TwiceDifferentiable(f, initial_x))
+    optimize(d, initial_x, method, options)
+end
```
Comment (Evizero) [inline src/optimize/multivariate/interface.jl]: why the two versions? ambiguities ?

Comment (Evizero) [inline src/optimize/multivariate/interface.jl]: (once with `f` and once with `f::Function`, while the rest stays the same)

**Comment by pkofod** [comment]
Seems like I need to look at this at v0.6 as well due to the type system changes

File: src/optimize/multivariate/interface.jl (hunk 1 above)
Comment (pkofod) [inline src/optimize/multivariate/interface.jl]: You mean why even have the `::Function` version and not leave it unspecified altogether?

Comment (Evizero) [inline src/optimize/multivariate/interface.jl]: yes. Just curious.

Comment (pkofod) [inline src/optimize/multivariate/interface.jl]: Don't really have a good answer. For some reason I just kept the `::Function` annotated ones, and added the ones with wider signatures. Removed them.

**Comment by pkofod** [comment]
`multivariate/interface.jl` should be *much* easier to read now (at least it is to me...)

**Pull request merged**